### PR TITLE
Improve thread safety of temp cred refresh

### PR DESCRIPTION
**Pull request by jamesls**
This builds on the pull request from https://github.com/boto/botocore/pull/764 and introduces a few additional enhancements for multithreaded credential refresh.  Includes:

* The original fix from #764.  Credentials must be accessed as a fixed set to ensure you don't get access_key from time t_1 and secret_key from time t_2.
* Introduce locking to ensure only one thread starts the refresh process.  The previous behavior was that every thread, once determining a refresh was needed, would all call the refresh function.  Additionally, double checked locking is used to avoid lock acquisition in the common case where a refresh isn't needed.
* Don't hard fail if refreshing initially fails.

To expand on the last point, there's now two time periods for refresh, the advisory period and the mandatory period.

##### Advisory period

* Only the thread that acquires the lock will refresh.  If a thread can't acquire the refresh lock during this period it uses the existing temporary creds.
* If the refresh function fails, a warning is logged, and the existing set of temporary creds are used.

##### Mandatory period

* All threads block on the refresh lock.
* Any exceptions raised during the refresh function will propagate back to the user.

There's still a few improvements that can be added (backoff delay during advisory period), but I think this is a good set of improvements.

Note: this builds on https://github.com/boto/botocore/pull/787 so the diff will show up here until that PR gets merged.

cc @kyleknap @rayluo @JordonPhillips

### CODE REVIEW

File: botocore/credentials.py [hunk 1]
```diff
@@ -255,27 +270,99 @@ def _seconds_remaining(self):
...
+                self._refresh_lock.release()
+        elif self.refresh_needed(self._mandatory_refresh_timeout):
+            # If we're within the mandatory refresh window,
+            # we must block until we get refreshed credentials.
+            with self._refresh_lock:
+                if not self.refresh_needed(self._mandatory_refresh_timeout):
+                    return
+                self._protected_refresh(mandatory=True)
+
+    def _protected_refresh(self, mandatory):
```
Comment (kyleknap) [inline botocore/credentials.py]: Ah so ``mandatory`` is suggesting if refresh is mandatory. Maybe call it ``is_mandatory``?

File: botocore/credentials.py [hunk 2]
```diff
@@ -289,6 +376,43 @@ def _set_from_data(self, data):
...
 
+    def get_frozen_credentials(self):
+        """Return immutable credentials.
+
+        The ``access_key``, ``secret_key``, and ``token`` properties
+        on this class will always check and refresh credentials if
+        needed before returning the particular credentials.
+
+        This has an edge case where you can get inconsistent
+        credentials.  Image this:
```
Comment (kyleknap) [inline botocore/credentials.py]: Image --> Imagine

File: tests/functional/test_credentials.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        threads = []
+        for _ in range(20):
+            threads.append(threading.Thread(target=func, args=(collected,)))
+        start = time.time()
+        for thread in threads:
+            thread.start()
+        for thread in threads:
+            thread.join()
+        for triple in collected:
+            self.assertTrue(triple[0] == triple[1] == triple[2], triple)
```
Comment (kyleknap) [inline tests/functional/test_credentials.py]: Might be good to explain what is in the triple and why you are checking that they are all equal. It is because we keep track of how many times refresh is called and the number of refresh times all need to be the same in order to ensure that the credentials set all came from the same refresh. Right?

File: tests/__init__.py [hunk 4]
```diff
@@ -242,3 +247,76 @@ def cmd(self, *cmd):
...
+            self._current_datetime() +
+            datetime.timedelta(seconds=creds_last_for))
+        if refresh_function is None:
+            refresh_function = self._do_refresh
+        super(IntegerRefresher, self).__init__(
+            '1', '1', '1', expires_in,
+            refresh_function, 'INTREFRESH')
+        self.creds_last_for = creds_last_for
+        self.refresh_counter = 0
+        if advisory_refresh is not None:
```
Comment (kyleknap) [inline tests/__init__.py]: What does ``None`` do for you here. You already have a default set. Not sure if the ``if`` check is needed.

**Comment by jamesls** [comment]
@kyleknap Feedback incorporated.

File: tests/functional/test_credentials.py (hunk 3 above)
Comment (jamesls) [inline tests/functional/test_credentials.py]: I added a comment here that explains what's going on, but yes, that's what we're verifying.

File: tests/__init__.py [hunk 5]
```diff
@@ -242,3 +247,74 @@ def cmd(self, *cmd):
...
+                 advisory_refresh=_advisory_refresh_timeout,
+                 mandatory_refresh=_mandatory_refresh_timeout,
+                 refresh_function=None):
+        expires_in = (
+            self._current_datetime() +
+            datetime.timedelta(seconds=creds_last_for))
+        if refresh_function is None:
+            refresh_function = self._do_refresh
+        super(IntegerRefresher, self).__init__(
+            '1', '1', '1', expires_in,
```
Comment (kyleknap) [inline tests/__init__.py]: Is their a particular reason you picked ``1`` as the starting point? It is a little confusing in the unit test that you assert that the value is ``2`` which I would see as the credentials being successfully refreshed twice when you are making sure that they get immediately refreshed. Maybe start at 0? Or have a helper function for asserting how many times credentials were refreshed?

File: tests/functional/test_credentials.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        # we should see a maximum number of refreshes being (6 / 2.0) + 1 = 4
+        max_calls_allowed = math.ceil((end - start) / 2.0) + 1
+        self.assertTrue(creds.refresh_counter <= max_calls_allowed,
+                        "Too many cred refreshes, max: %s, actual: %s, "
+                        "time_delta: %.4f" % (max_calls_allowed,
+                                              creds.refresh_counter,
+                                              (end - start)))
+
+    def test_no_race_for_immediate_expiration(self):
+        creds = IntegerRefresher(
```
Comment (kyleknap) [inline tests/functional/test_credentials.py]: How does this setup for immediate expiration? If the creds are set to last for 1 sec and the advisory period is 1 sec while the mandatory refresh is at 0 sec, should the credentials still be in the advisory period? Or is there a clock tick somewhere I am missing in my logic?

**Comment by kyleknap** [comment]
I only had a couple questions about the tests that be cool to get addressed. Otherwise looks good, :ship:

File: tests/__init__.py (hunk 5 above)
Comment (jamesls) [inline tests/__init__.py]: I think starting at 0 makes sense.  That way the value of akid/sak/token is the number of refreshes that have occurred.

File: tests/functional/test_credentials.py (hunk 6 above)
Comment (jamesls) [inline tests/functional/test_credentials.py]: I can rename this, but it was immediate "advisory" expiration, that is, the credentials need to be refreshed immediately after they're refreshed.

**Comment by kyleknap** [comment]
Thanks. :ship:
